src/middleware/jira-symmetric-jwt-middleware.ts

```diff
@@ -0,0 +1,137 @@
+			return res.status(401).send("Unauthorised");
+		}
+
+		const installation = await Installation.getForClientKey(issuer);
+		if (!installation) {
+			req.log.warn("No Installation found");
+			return res.status(401).send("Unauthorised");
+		}
+
+		const secret = await installation.decrypt("encryptedSharedSecret");
+
+		const tokenType = checkPathValidity(req.originalUrl) && req.method == "GET" ? TokenType.normal : TokenType.context;
+		try {
+			verifySymmetricJwt(token, secret, req, tokenType, req.log);
+		} catch (err) {
+			req.log.warn({ err }, "Could not verify symmetric JWT");
+			return res.status(401).send("Unauthorised");
+		}
+
+		res.locals.installation = installation;
+		res.locals.jiraHost = installation.jiraHost;
+		req.session.jiraHost = installation.jiraHost;
+
+		if (req.cookies.jwt) {
+			res.clearCookie("jwt");
+		}
+		return next();
+
+	} else if (req.session?.jiraHost) {
+
+		const installation = await Installation.getForHost(req.session.jiraHost);
+		if (!installation) {
+			req.log.warn("No Installation found");
+			return res.status(401).send("Unauthorised");
+		}
+
+		res.locals.installation = installation;
+		res.locals.jiraHost = installation.jiraHost;
+		return next();
+	}
+
+	req.log.warn("No token found and session cookie has not jiraHost");
+	return res.status(401).send("Unauthorised");
+
+};
+
+const getIssuer = (token: string, logger: Logger): string | undefined => {
+
+	let unverifiedClaims;
+	try {
+		unverifiedClaims = decodeSymmetric(token, "", getAlgorithm(token), true); // decode without verification;
+	} catch (err) {
+		logger.warn({ err }, "Invalid JWT");
+		throw new Error(`Invalid JWT: ${err.message}`);
+	}
+
+	if (!unverifiedClaims.iss) {
+		throw new Error("JWT claim did not contain the issuer (iss) claim");
+	}
+
+	return unverifiedClaims.iss;
+};
+
+const verifySymmetricJwt = (token: string, secret: string, req: Request, tokenType: TokenType, logger: Logger): boolean => {
+	const algorithm = getAlgorithm(token);
+
+	/* eslint-disable @typescript-eslint/no-explicit-any*/
+	let verifiedClaims: any; //due to decodeSymmetric return any
+	try {
+		verifiedClaims = decodeSymmetric(token, secret, algorithm, false);
+	} catch (err) {
+		logger.warn({ err }, "Invalid JWT");
+		throw new Error(`Unable to decode JWT token: ${err.message}`);
+	}
+
+	return verifyJwtClaims(verifiedClaims, tokenType, req);
+};
+
+export const verifyJwtClaims = (verifiedClaims: { exp: number, qsh: string }, tokenType: TokenType, req: Request): boolean => {
+	const expiry = verifiedClaims.exp;
+
+	if (expiry && (Date.now() / 1000 >= expiry)) {
+		throw new Error("JWT Verification Failed, token is expired");
+	}
+
+	if (verifiedClaims.qsh) {
+		let qshVerified: boolean;
+		if (tokenType === TokenType.context) {
+			//If we use context jsw tokens, their qsh will be constant
+			qshVerified = verifiedClaims.qsh === "context-qsh";
+		} else {
+			//validate query string hash
+			qshVerified = verifyQsh(verifiedClaims.qsh, req);
```

Comment:
do we have a test for this, when it is not valid?

Reply:
test case added